Added bindings for Js.Re module

glennsl opened the pull request.

bobzhang commented on jscomp/others/js_re.ml:
```diff
@@ -40,9 +40,26 @@
 
 type t
+type result = string array
+(* `result` also has `index` and `input` properties, but don't know how to represent that *)
```
can you type `result` as abstract type and type it as an array like object instead?

```ocaml
type result 
external matches : result -> string array ..
external index : result -> ...
```

glennsl replied:
Great idea!

bobzhang commented on jscomp/others/js_re.mli:
```diff
@@ -40,9 +40,25 @@
 type t
+type result = string array
+
+val fromString : string -> t
```
we need keep `external` instead of `val`, however, if it is  all about types, we only need `mli` file though

glennsl replied:
Nice as long as the `.ml` can be removed, but why not `val` if there must be an `.ml`? It seemed to work, and the OCaml manual says it's ok.

glennsl commented:
Changes made according to your comments.

bobzhang commented:
@glennsl thanks, keeping externals so that compiler can do more inlining, otherwise they are the same. would you adapt your commit message based on https://github.com/bloomberg/bucklescript/blob/master/CONTRIBUTING.md
I will merge it once it is finished

bobzhang commented on jscomp/others/js_re.mli:
```diff
@@ -17,12 +17,12 @@
  * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)
 (* Copyright (C) 2015-2016 Bloomberg Finance L.P.
- * 
+ *
```
there seems to be two copy right headers, would you do a clean up

bobzhang commented on jscomp/others/js_re.mli:
```diff
@@ -40,9 +40,29 @@
+external source : t -> string = "" [@@bs.get]
+external sticky : t -> Js.boolean = "" [@@bs.get]
+external unicode : t -> Js.boolean = "" [@@bs.get]
+
```
@glennsl do you think it makes sense to provide some sugar for such FFI as I proposed here: https://github.com/bloomberg/bucklescript/blob/master/CONTRIBUTING.md

glennsl replied:
Wrong link. Is this what you mean?
```
external sticky : t -> Js.boolean = "" [@@bs.get]
let sticky: t -> bool = fun self  -> Js.to_bool (sticky self)
```
If so then yes, I would have if it didn't seem like using Js.boolean was the convention.

bobzhang replied:
ah, I am fine with `Js.boolean`, this link https://github.com/bloomberg/bucklescript/issues/522

glennsl replied:
Oh, and it's not been added yet, you're just asking my opinion? It would certianly be nice for writing lots of bindings, but it's also more syntax to learn and will make it harder for beignners to read. So I think it should be avoided unless there's good documentation.

glennsl commented:
Thanks for explaining, I should have guessed it was performance related ;)

Is adding a Sign-off to the PR description OK? Or do I need to rebase and force push? to alter each commit message?

bobzhang commented:
@glennsl it would be nice to have a rebase with a single commit ; )

glennsl commented:
Right, I'll do my best, but I'm not a git wizard :)

glennsl commented:
Removed the duplicate copyright header, added more tests, squashed the commits and added sign-off. Learned that Travis uses Node v4, so disabled some tests for unsupported es2015 functionality. I think that should do it.

bobzhang commented:
thanks, we can upgrade to v6 for tests though
